hypothesis-python/src/hypothesis/internal/escalation.py

```diff
@@ -158,3 +158,19 @@ def format_exception(err, tb):
 
     # If all else fails, use the standard-library formatting tools
     return "".join(traceback.format_exception(type(err), err, tb))
+
+
+def _find_user_stack_level() -> int:
```

Comment:
@Zac-HD I want to open a thread specifically to talk about this proposed change, since I think it could be controversial.

**support for inspecting the stack like this**

* ensures that warnings continue to point to user code correctly even if future refactorings of `hypothesis` re-arrange things
* handles the case "several different code paths can end up calling the same `hypothesis` code" without needing to add a `stacklevel` argument to any `hypothesis` APIs and pass it around through calls

**reasons this might be a bad idea**

* more computational overhead than passing around integer literals for `stacklevel`
* maybe there are cases where the call stack size can be large enough that this could cause significant memory pressure
    - reading through the `traceback` docs ([link](https://docs.python.org/3/library/traceback.html)) and source code ([link](https://github.com/python/cpython/blob/46fc584b00f2ccee5b77d8fc525881def8670a02/Lib/traceback.py)), I don't see a way to iterate over the stack from newest to oldest ([traceback.extract_stack() source code](https://github.com/python/cpython/blob/46fc584b00f2ccee5b77d8fc525881def8670a02/Lib/traceback.py#L220-L233))...so I think you have to pull the whole thing into memory and reverse it
* maybe this doesn't play well with async code (I'm not sure how those libraries like `asyncio` interact with the stack)


Thanks for talking through it with me 😁

Reply:
I also won't work on fixing the failing `check-coverage` job ([build link](https://github.com/HypothesisWorks/hypothesis/runs/7378705366?check_suite_focus=true)) until this thread is resolved.

Reply:
As you suspect, I'm not a fan of this code sorry - among other issues, is only detects the stack depth on editable installs (and fixing that would leave zip installs, use in pyodide, etc.)

Ugly as passing a `stacklevel=` argument through the array API might be, I think it's the best solution here.  If you want to skip that and just implement the two constant-stacklevel parts, I'm happy to defer the last warning to a future PR.

Reply:
> I'm not a fan of this code

no prob! Had a feeling this was possibly controversial. Thanks for talking through it with me. I'll revert those changes.

> I explicitly judged that the warnings not listed at the top of the issue are not worth fixing

Ok I see. That wasn't obvious to me from the issue and I thought that maybe they'd just been missed when writing it up. I won't make any changes to warnings not mentioned in the tope of #3403.